cloud/store/client.go

```diff
@@ -30,63 +35,114 @@ func NewClient(localService Cluster) *Client {
 	return &Client{localService: localService}
 }
 
-func (cl Client) Join(leaderAddr string, req *cmd.JoinPeerRequest) error {
-	log.Printf("client join: %s %+v\n", leaderAddr, req)
-	if leaderAddr == cl.localService.address {
-		err := cl.localService.Join(req.Id, req.Address, req.Voter)
-		if err == nil {
-			return err
+// Join joins this node to an existing cluster identified by its leader's address.
+// If a new leader has been elected, the request is redirected to the new leader.
+func (cl *Client) Join(ctx context.Context, leaderAddr string, req *cmd.JoinPeerRequest) (*cmd.JoinPeerResponse, error) {
+	resp, err := cl.join(ctx, leaderAddr, req)
+	if err != nil {
+		st := status.Convert(err)
+		if leader := resp.GetLeader(); st.Code() == codes.NotFound && leader != "" {
```

Comment:
We should have a config parameter `RAFT_LEADER_PROXY_RETRIES` and make the leader proxy code retry this amount of times the request if it fails due to no leader reached. Same for other RPC request that should be directed to the leader.

Reply:
I don't believe that retries would be effective in this scenario. If the target node is either the leader or knows the leader, then the request is successful. Otherwise, we must attempt the request with another node. This logic is employed in the bootstrapper

Reply:
I had in mind the case where might reach nodeA that we believe is the leader and it no longer is. But you're right that then the node will forward to the new leader. So it should work out.